crypto/src/key/hdkd/child_number.rs

```diff
@@ -98,6 +149,68 @@ impl fmt::Display for ChildNumber {
                 let alt = f.alternate();
                 f.write_char(if alt { HARDENED_H } else { HARDENED_APOS })
             }
+            DerivationType::Normal(index) => fmt::Display::fmt(&index, f),
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use crate::key::hdkd::child_number::ChildNumber;
+    use crate::key::hdkd::u31;
+    use crate::key::hdkd::u31::U31;
+    use rstest::rstest;
+    use serialization::{DecodeAll, Encode};
+    use std::str::FromStr;
+
+    fn examine_child_number(num: ChildNumber, encoded_num: u32, is_hardened: bool) {
+        assert_eq!(num.is_normal(), !is_hardened);
+        assert_eq!(num.is_hardened(), is_hardened);
+        assert_eq!(num.into_encoded_index(), encoded_num);
+        assert_eq!(num.into_encoded_be_bytes(), encoded_num.to_be_bytes());
+        assert_eq!(
+            num,
+            ChildNumber::decode_all(&mut num.encode().as_slice()).unwrap()
+        );
+    }
+
+    #[rstest]
+    #[trace]
+    #[case(0, false)]
+    #[case(1, false)]
+    #[case(1234567, false)]
+    #[case(u32::MAX & (!0x80000000 - 1), false)]
+    #[case(u32::MAX & !0x80000000, false)]
+    #[case(u32::MAX & (!0x80000000 + 1), true)]
+    #[case(u32::MAX - 1, true)]
+    #[case(u32::MAX, true)]
+    fn create_child_number(#[case] encoded_num: u32, #[case] is_hardened: bool) {
```

Comment:
Please another test where a random value is encoded, decoded, and then we ensure that it's the same. The random value should be treated in the same way we do our property-based testing (like we discussed the other day).